ref-arch/real-time-analytics.rst

```diff
@@ -0,0 +1,509 @@
+this document we'll only talk about the first granularity, the 1-minute one. The github
+repo has `DDL <http://github.com>`_ for the other resolutions.
+
+.. code-block:: sql
+
+  CREATE TABLE http_request_1min (
+        zone_id INT,
+        ingest_time TIMESTAMPTZ, -- which minute this row represents
+
+        error_count INT,
+        success_count INT,
+        request_count INT,
+        average_response_time_msec INT,
+        CHECK (request_count = error_count + success_count)
+  )
+  SELECT master_create_distributed_table('http_requests_1min', 'zone_id', 'hash');
+  SELECT master_create_worker_shards('http_requests_1min', 16, 2);
+  
+  -- indexes aren't automatically created by Citus
+  -- this will create the index on all shards
+  CREATE INDEX ON http_requests_1min (zone_id, ingest_time);
+
+This looks a lot like the previous code block. Most importantly: It also shards on
+``zone_id``, and it also uses 16 shards with 2 replicas of each. Because all three of
+those match, there's a 1-to-1 correspondence between ``http_request`` shards and
+``http_request_1min`` shards, and Citus will place matching shards on the same worker.
+This is called colocation; it makes queries such as joins faster and our rollups possible.
+
+.. image:: /images/colocation.png
+  :alt: colocation in citus
+
+In order to populate ``http_request_1min`` we're going to periodically run the equivalent
+of an INSERT INTO SELECT. Citus doesn't yet support INSERT INTO SELECT on distributed
+tables, so instead we'll run a function on all the workers which runs INSERT INTO SELECT
+on every matching pair of shards (this is possible because the shards are colocated)
+
+.. code-block:: plpgsql
+
+  -- this should be run on each worker
+  CREATE FUNCTION rollup_1min(source_shard text, dest_shard text) RETURNS void
+  AS $$
+  DECLARE
+    v_latest_minute_already_aggregated timestamptz;
+    v_new_latest_already_aggregated timestamptz;
+  BEGIN
+    PERFORM SET lock_timeout 100;
+    -- since master calls this function every minute, and future invokations will
+    -- do any work this function doesn't do, it's safe to quit if we wait too long
+    -- for this FOR UPDATE lock which makes sure at most one instance of this function
+    -- runs at a time
+    SELECT ingest_time INTO v_latest_minute_already_aggregated FROM rollup_thresholds
+      WHERE source_shard = source_shard AND dest_shard = dest_shard
+      FOR UPDATE;
+    IF NOT FOUND THEN
+      INSERT INTO rollup_thresholds VALUES (
+        '1970-01-01', source_shard::regclass, dest_shard::regclass);
+      RETURN;
+    END IF;
+    PERFORM RESET lock_timeout;
+
+    EXECUTE format('
+      WITH (
+        INSERT INTO %I (
+            zone_id, ingest_time, request_count,
+            error_count, success_count, average_response_time_msec)
+          SELECT
+            zone_id,
+            date_trunc('minute', ingest_time) as minute,
```

Comment:
The single quotes should be quoted: `''minute''` since it's in a string for format().